File: 04-launch-darkly/node.js/src/data/feature-flag-repository.ts
import * as ld from '@launchdarkly/node-server-sdk';
import { FeatureFlags } from '../types/feature-flag.js';


/*
LaunchDarkly SDK docs
https://docs.launchdarkly.com/sdk/server-side/node-js
*/

const LAUNCH_DARKLY_SDK_KEY = process.env.LAUNCH_DARKLY_SDK_KEY;
if (!LAUNCH_DARKLY_SDK_KEY) {
  throw new Error('set LAUNCH_DARKLY_SDK_KEY environment variable');
}
const client = ld.init(LAUNCH_DARKLY_SDK_KEY);


export default async function getFlag(key: FeatureFlags, name?: string | undefined, defaultValue: boolean = false): Promise<boolean> {
  // A context is a person, service, or machine
  // https://docs.launchdarkly.com/sdk/features/context-config
  const context = {
    kind: 'user',
    key: 'demo',
    name: name,
  };

  const value = await client.variation(key, context, defaultValue);
  return value;
}

export async function getAllFlags(name?: string | undefined): Promise<{ [key: string]: boolean }> {
  const context = {
    kind: 'user',
    key: 'demo',
    name: name,
  };

  const flags = await client.allFlagsState(context);
  const json = flags.toJSON() as { [key: string]: boolean };
  const keys = Object.keys(json);
  keys.filter(k => k.startsWith('$')).forEach(k => delete json[k]);
  return json;
}

File: 05-launch-darkly/dotnet/Data/FeatureFlagRepository.cs
namespace FeatureFlagLaunchDarkly.Data;

/*
Launch Darkly docs:
- https://docs.launchdarkly.com/home/getting-started
- https://docs.launchdarkly.com/sdk/server-side/dotnet
*/

public interface IFeatureFlagRepository
{
    bool FeatureIsEnabled(FeatureFlags featureFlag, string? user = null, bool defaultValue = false);
    Dictionary<string, bool> GetAllFeatureFlags(string? user = null);
}

public class FeatureFlagRepository(ILdClient launchDarklyClient) : IFeatureFlagRepository
{
    public bool FeatureIsEnabled(FeatureFlags featureFlag, string? user = null, bool defaultValue = false)
    {
        // A context is a person, service, or machine
        // https://docs.launchdarkly.com/sdk/features/context-config
        var context = Context.Builder("demo")
            .Kind("user")
            .Name(user ?? "")
            .Build();

        bool flagValue = launchDarklyClient.BoolVariation(featureFlag.ToString(), context, defaultValue);
        return flagValue;
    }

    public Dictionary<string, bool> GetAllFeatureFlags(string? user = null)
    {
        var context = Context.Builder("demo")
            .Kind("user")
            .Name(user ?? "")
            .Build();

        var state = launchDarklyClient.AllFlagsState(context);
        if (state == null)
        {
            return new Dictionary<string, bool>();
        }
        var values = state.ToValuesJsonMap();

        return values.ToDictionary(v => v.Key, v => v.Value.AsBool);
    }
}
